src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm

```diff
@@ -0,0 +1,238 @@
+/**
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "MTRDeviceControllerDataStore.h"
+#import "MTRLogging_Internal.h"
+
+// TODO: FIXME: Figure out whether these are good key strings.
+static NSString * sResumptionNodeListKey = @"caseResumptionNodeList";
+static NSString * sLastUsedNOCKey = @"lastUsedControllerNOC";
+
+static NSString * ResumptionByNodeIDKey(NSNumber * nodeID)
+{
+    return [NSString stringWithFormat:@"caseResumptionByNodeID/%llx", nodeID.unsignedLongLongValue];
+}
+
+static NSString * ResumptionByResumptionIDKey(NSData * resumptionID)
+{
+    return
+        [NSString stringWithFormat:@"caseResumptionByResumptionID/%s", [resumptionID base64EncodedStringWithOptions:0].UTF8String];
+}
+
+@implementation MTRDeviceControllerDataStore {
+    id<MTRDeviceControllerStorageDelegate> _storageDelegate;
+    dispatch_queue_t _storageDelegateQueue;
+    MTRDeviceController * _controller;
+    // Array of nodes with resumption info, oldest-stored first.
+    NSMutableArray<NSNumber *> * _nodesWithResumptionInfo;
+}
+
+- (instancetype)initWithController:(MTRDeviceController *)controller
+                   storageDelegate:(id<MTRDeviceControllerStorageDelegate>)storageDelegate
+              storageDelegateQueue:(dispatch_queue_t)storageDelegateQueue
+{
+    if (!(self = [super init])) {
+        return nil;
+    }
+
+    _controller = controller;
+    _storageDelegate = storageDelegate;
+    _storageDelegateQueue = storageDelegateQueue;
+
+    __block id resumptionNodeList;
+    dispatch_sync(_storageDelegateQueue, ^{
+        resumptionNodeList = [_storageDelegate controller:_controller
+                                              valueForKey:sResumptionNodeListKey
+                                            securityLevel:MTRStorageSecurityLevelSecure
+                                              sharingType:MTRStorageSharingTypeSameIdentityAllowed];
+    });
+    if (resumptionNodeList != nil) {
+        if (![resumptionNodeList isKindOfClass:[NSMutableArray class]]) {
```

Comment:
Couldn't this be an NSArray when unarchived?

Reply:
Or a CFType as well?

Reply:
> Couldn't this be an NSArray when unarchived?

We pass an NSMutableArray to the consumer as an opaque id.  They archive it.  They're not supposed to care that it's an array, just archive it as-is.  When they unarchive, how would it end up as something other than NSMutableArray?

Reply:
When you archive/encode a NSMutableArray, the unarchive/decode would return the same mutable array as well.

Reply:
Not necessarily FYI.